client/src/components/survey-edit/survey-edit-card.js

```diff
@@ -0,0 +1,241 @@
+    this.doEditEnd = this.doEditEnd.bind(this)
+    this.doEditCancel = this.doEditCancel.bind(this)
+    this.handleTextEdit = this.handleTextEdit.bind(this)
+    this.handleKeyPress = this.handleKeyPress.bind(this)
+  }
+
+  doOpen() {
+    const { onOpen, question } = this.props
+    onOpen(question.id)
+  }
+
+  doClose() {
+    const { onClose, question } = this.props
+    onClose(question.id)
+  }
+
+  doDelete() {
+    const { onDelete, question } = this.props
+    onDelete(question.id)
+  }
+
+  doEditBegin() {
+    const { onEditBegin, question } = this.props
+    onEditBegin(question.id)
+  }
+
+  doEditEnd() {
+    const { onEditEnd, question } = this.props
+    onEditEnd(question.id)
+  }
+
+  doEditCancel() {
+    const { onEditCancel, question } = this.props
+    onEditCancel(question.id)
+  }
+
+  handleTextEdit(event) {
+    const { onEditChange, question } = this.props
+    const text = event.target.value.replace(/[\r\n\t]/g, '')
+    onEditChange(question.id, { text })
+  }
+
+  handleKeyPress(event) {
+    if (event.keyCode === 13) {
+      // Enter: End edit
+      this.doEditEnd()
+    }
+  }
+
+  render() {
+    const { question: originalQuestion, edit, onEditChange, opened, busy } = this.props
+
+    const edited = edit.id === originalQuestion.id
+    const question = edited ? R.mergeDeepRight(originalQuestion, edit) : originalQuestion
+
+    const answers = question.answers || []
+    const commit = edited && edit.commit
+
+    const dataDisplay = {
+      text: () => (<TextDisplay question={question} onOpen={this.doOpen} />),
+      option: () => (<OptionDisplay question={question} />),
+      info: () => (<div />),
+    }
+
+    const dataEdit = {
+      text: () => (<div />),
+      option: () => (
+        <OptionEdit
+          question={question}
+          onEditChange={onEditChange}
+          onKeyPress={this.handleKeyPress}
+          commit={commit}
+        />
+      ),
+      info: () => (<div />),
+    }
+
+    const display = edited ? dataEdit[question.type]() : dataDisplay[question.type]()
+
+    return (
+      <div className={css('survey-edit-card', { 'edit-commit': commit })}>
+
+        <div className={css('header')}>
+
+          {edited ? (
+            <>
+              <textarea
+                value={question.text}
+                onChange={this.handleTextEdit}
+                placeholder="Write a question..."
+                autoFocus
+                onKeyPress={this.handleKeyPress}
+                disabled={commit}
+              />
+              <button
+                type="button"
+                onClick={this.doEditEnd}
+                tabIndex="-1"
+                disabled={commit}
+              >OK
+              </button>
+              <button
+                type="button"
+                onClick={this.doEditCancel}
+                tabIndex="-1"
+                disabled={commit}
+              >Cancel
+              </button>
+            </>
+          ) : !opened ? (
```

Comment:
quite complicated tertiary clause, is it still fine? your call

Reply:
Could these be moved to functions to make code a bit clearer?

Reply:
solved

Reply:
do nothing, ship it